### ISSUE TITLE
T28: Run compact additive greedy best-first development cells

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/multimodality_on_planning#66]
> *This was generated by AI during triage and revised for the issue-64 VLM-feasible v3 corpus.*

**Parent:** #38
**Plan ticket:** T28

## What to build
Run the compact additive greedy best-first development conditions for `best_first_add_greedy`: pretrained base, process-SFT, random-valid, and exact-reference on the matched issue-64 v3 development panel, cell-for-cell with #65.

The declared scalar priority is `h_add` with canonical generation-serial tie-breaking and no closed-node reopening. Plan optimality is not claimed or required.

## Acceptance criteria
- All four development conditions use the same issue-64 v3 selected development tasks and are cell-for-cell matched to #65.
- Each selected exact reference contains at most 1,024 decisions; the frozen 2x allowance permits at most 2,048 VLM calls per episode.
- Qualify actual VLM throughput and wall-clock feasibility before rollout; issue a governed stop instead of silently reducing or segmenting an authorized episode.
- Run exactly one training seed (`17`) for each distinct model/cell configuration; rollout/reference seeds do not imply training-seed replication.
- The active path uses no hashes, checksums, artifact comparisons, regeneration comparisons, or integrity-check mode. Checks are limited to scientific episode semantics, algorithm invariants, coverage, and resource limits.
- No run starts without the matching authorization and gate receipt; outcomes distinguish `PASS`, `VALID_STOP`, `INVALID`, and `ANCESTOR_STOP`, and only `PASS` is scientific completion.

## Dependencies
#52 and #64 are complete.

---
**Governed vocabulary reference:** This ticket uses the governed vocabulary (operational competence, structural/process competence, validity, stop outcome, text-state, visual-state, multimodal-state) exactly as defined in `CONTEXT.md` section "Governed downstream vocabulary", governed by issue #39 (T01). Terms are reused verbatim; no local redefinition applies.

**Comment by Sino-Huang** [comment]
Apply the resource-bounded evaluation protocol established by #54:

- drive incremental episode sessions in deterministic rounds and use batched Transformers generation with one shared backbone per GPU, adapter switching, adapter-isolated caching, and at most one request per active episode per round;
- freeze a 2x matching-reference decision-call limit separately from expansion limits; exhaustion is a valid unsuccessful episode;
- evaluate only final checkpoints across five seeds, with base, learned, random-valid, and exact-reference conditions on the identical selected tasks;
- qualify throughput without success outcomes before rollout, using the actual precision/batch settings and a cost-qualified fallback panel if full coverage cannot fit;
- stop before new launches at the wall-clock cutoff, preserve completed episodes, replay every completed episode, and never let partial coverage satisfy the gate.

Report invariant-valid success, invalid-operation rate, budget usage, and learned-vs-best-control gain separately. #54 reached 100% learned success on its reduced panel but still correctly `VALID_STOP` because random-valid also reached 100%; avoid interpreting panel success alone as structural advantage.

**Comment by Sino-Huang** [comment]
Supervisor time-budget clarification: this ticket must not require multiple independent training-seed replicates. For each distinct model/cell configuration, authorize exactly one training run using seed 17 and evaluate only the final checkpoint from that run.

Any existing five-seeds requirement must be interpreted as rollout/reference/evaluation seeds applied to the single trained checkpoint where repeated evaluation is still useful, not as five separately trained checkpoints. Do not estimate or claim training-seed variance, and do not substitute rollout-seed variation for it; use whole-problem-instance uncertainty where applicable.

If a frozen scientific criterion truly requires independent training-seed replication, issue the appropriate governed VALID_STOP or gated-scope receipt instead of launching additional training runs.

**Comment by Sino-Huang** [comment]
Replacement selection after the retained issue-62 optimal-A* `VALID_STOP`:

- Independent Fast Downward screening on all 25 fixed development tasks found `eager_wastar([add()],w=2)` solved 25/25 with 16,622 total expansions and aggregate plan cost 1,699; `eager_greedy([add()])` without preferred operators solved 25/25 with 9,542 total expansions and aggregate plan cost 2,098.
- LM-cut/landmark variants were rejected: weighted LM-cut timed out on the hard VisitAll task even with few expansions because evaluation itself was expensive.
- Intermediate additive weights 3 and 5 lay between the selected endpoints and did not add a distinct scientific condition.
- The production adapters are named `best_first_add_w2` and `best_first_add_greedy`. Both use one eager scalar-priority runtime and `h_add`; the only setting differences are the priority formula and the standard reopening rule.
- The compact trace contract persists one copy of task/state facts plus per-decision typed target, trusted runtime result, and exact-input digest. Independent replay reconstructs every authoritative candidate/pruning fact and byte-identical input. Full frontier arrays, repeated complete inputs, landmarks, preferred operators, multiple queues, lazy evaluation, and trace segmentation are excluded.

This is a deliberately suboptimal best-first study. It makes no optimal-plan or admissible-search claim.

**Comment by Sino-Huang** [comment]
Issue #64 is complete. This ticket is now bound cell-for-cell to #65 on the 64-pair v3 feasibility panel: maximum 1,024 reference decisions and 2,048 allowed VLM calls per episode, with throughput qualification required before launch. No integrity mechanism is authorized.

**Comment by Sino-Huang** [comment]
Implementation is pushed in `3ca5cc1` (`feat(planning): add greedy best-first development runner`).

Implemented for contract `issue-66-best-first-add-greedy-development-v1`:
- the same 23 issue-64 v3 development tasks as #65, using only `best_first_add_greedy`;
- 6,696 exact development decisions, observed maximum 923, and the frozen separate 2x per-episode call allowance (maximum 1,846);
- one two-epoch process-SFT LoRA run with training seed 17 only, 8,342 train rows, 6,696 dev rows, and 522 optimizer steps;
- exact and five-seed random-valid references plus five rollout seeds for pretrained base and the single learned checkpoint;
- outcome-blind two-A100 float32 qualification before training or rollout, complete-panel-only qualification, two cost-balanced GPU rollout shards, semantic replay, and governed adjudication;
- terminal progress and ETA logs, including 30-second heartbeats during model loading and blocking generation batches;
- distinct issue-66 authorization, schemas, receipt IDs, output root, dataset root, and explicit default `MASTER_PORT=29660`.

The active path creates no hashes, checksums, artifact comparisons, regeneration comparisons, or follow-up check mode.

Validation completed without a GPU/model run:
- the local greedy SFT projection materialized successfully: 8,342 train / 6,696 dev rows (62 MB);
- exact and random-valid greedy episodes passed semantic execution and replay probes;
- complete planning benchmark: 361 passed (one unrelated Albumentations update warning);
- Ruff, Black, and basedpyright pass;
- the complete workflow dry-run passes and leaves the experiment output root clean.

Operator command:
```bash
source ~/cd_vlaplan
cd /data/scratch/projects/punim0478/sukaih/multimodality_on_planning
python scripts/run_best_first_issue66.py all \
  --devices cuda:0 cuda:1 \
  --reference-workers 8 \
  --training-device 1 \
  --master-port 29660
```

The command qualifies first. If qualification cannot certify the complete product inside the frozen 15-hour rollout budget, it records and adjudicates `VALID_STOP` without starting training. If it passes, it continues through preparation, references, training, rollout, and adjudication. Every long stage prints live progress. Resume an interrupted pre-cutoff run by adding `--resume`. Detailed stage commands are in `docs/issue-66-best-first-add-greedy-development.md`.

The ticket remains open pending the real qualification and, only if authorized by that result, the remaining experiment.

**Comment by Sino-Huang** [comment]
Issue #66 completed under contract `issue-66-best-first-add-greedy-development-v1` with scientific `PASS`. The compact terminal result is retained in `22bab14` at `data/best_first_paired_phase_v3/issue66-terminal/result.json`.

Completion evidence:
- outcome-blind two-A100 qualification certified the complete 23-task issue-64 v3 development panel; lower-95% throughput was 0.372470 calls/s and projected rollout time was 43,248.16 seconds, below the frozen 15-hour ceiling;
- the selected traces remained bounded at a maximum of 923 reference decisions and 1,846 calls per episode;
- exactly one training run was performed: seed 17, two epochs, 522/522 optimizer steps, final checkpoint `checkpoint-522`;
- references completed: 23 exact-reference and 115 random-valid episodes;
- model rollout completed: 115 pretrained-base and 115 process-SFT episodes across two balanced GPU shards;
- all 368 retained episodes were independently reconstructed through the trusted runtime with the same scientific inputs, transitions, and results;
- final gate receipt: `gate:issue-66-best-first-add-greedy-development-v1:attempt-001`, `PASS`, `scientific_completion: true`.

Results:
- exact reference: 23/23 invariant-valid successes;
- pretrained base: 0/115; all 115 terminated after one deterministic invalid operation;
- process SFT: 115/115, zero invalid operations, 0.5 call-budget usage;
- random-valid: 115/115, zero invalid operations;
- learned gain over the best control: 0.0; paired whole-instance bootstrap lower bound: 0.0.

The bounded interpretation is therefore: process SFT completely repaired the pretrained base operational failure and executed every development episode correctly, but it did not establish superiority over random-valid because that control also saturated the panel. This is a completed development result, not a positive learned-versus-best-control advantage claim.

The active run used scientific semantic validation only and introduced no artifact-comparison or regeneration mechanism.

**Comment by Sino-Huang** [comment]
Interpretation clarification: `random_valid` is a programmatic, oracle-assisted reference policy, not a comparable learned-model baseline. It reads the declared structured successor-candidate set and emits a well-formed valid operation by construction, so it does not face the VLMs parsing and invalid-output failure modes.

The positive result of #66 is operational search competence: process SFT moved from the pretrained bases 0/115 completed episodes to 115/115 invariant-valid completed episodes, with zero invalid operations and no expert repair during evaluation. Parity with random-valid limits only the stronger claim that the model learned a superior search strategy. It does not negate the result that SFT taught the model to execute the exposed search protocol reliably. The ticket remains correctly completed and closed.
